File: Python/problem0070.py
# -*- coding: utf-8 -*-
"""
Created on Thu May 23 16:05:33 2019

@author: Administrator
"""

class Solution:
    def climbStairs(self, n: int) -> int:
        if n == 1:
            return 1
        if n == 2:
            return 2
        a = {}
        a[1] = 1
        a[2] = 2
        for i in range(3, n+1):
            a[i] = a[i-1] + a[i-2]
        return a[n]
        
#        if n not in a.keys():
#            a[n] = self.climbStairs(n-1) + self.climbStairs(n-2)
#        return a[n]
        
solu = Solution()
n = 2
n = 3
print(solu.climbStairs(n))

File: C++/problem0070.cpp
//代码中使用的 \texttt{pow}pow 函数的时空复杂度与 CPU 支持的指令集相关；

class Solution {
public:
    int climbStairs(int n) {
        // if (n == 1)
        // {
        //     return 1;
        // }
        // else if (n == 2)
        // {
        //     return 2;
        // }
        // return climbStairs(n - 1) + climbStairs(n - 2);

        vector<int> v{1, 2};
        for (int i = 2; i < n; ++i)
        {
            v.push_back(v[i - 1] + v[i - 2]);
        }
        // cout << n - 1 << ' ' << v.size() << endl;
        return v[n - 1];
        // return v[v.size() - 1];
    }
};
